resolver/src/p2p/fetcher.rs

```diff
@@ -75,146 +126,247 @@ pub struct Fetcher<E: Clock + Rng + Metrics, P: PublicKey, Key: Span, NetS: Send
-                },
-                self.priority_requests,
-            )
-            .await;
-        let result = match result {
-            Err(err) => Err(SendError::Failed::<NetS>(err)),
-            Ok(to) if to.is_empty() => Err(SendError::Empty),
-            Ok(_) => Ok(()),
-        };
-
-        // Insert the request into the relevant map
-        match result {
-            // If the message was not sent successfully, treat it instantly as a peer timeout
-            Err(err) => {
-                debug!(?err, ?peer, "send failed");
-                let req = self.requester.handle(&peer, id).unwrap(); // Unwrap is safe
-                self.requester.fail(req);
-                self.add_retry(key);
+                };
+                match checked.send(message, self.priority_requests).await {
+                    Ok(sent) if !sent.is_empty() => {
+                        // Success - remove from pending and add to active
+                        self.requests_sent.inc(Status::Success);
+                        self.pending.remove(&key);
+                        self.active.put(id, deadline);
+                        self.requests.insert(
+                            id,
+                            ActiveRequest {
+                                key: key.clone(),
+                                peer,
+                                start: now,
+                            },
+                        );
+                        self.key_to_id.insert(key, id);
+                        return;
+                    }
+                    Ok(_) => {
+                        // Empty result - peer dropped message, try next
+                        all_rate_limited = false;
+                        self.requests_sent.inc(Status::Dropped);
+                        debug!(?peer, "send returned empty");
+                        continue;
+                    }
+                    Err(err) => {
+                        // Send error - not rate-limited, just failed
+                        all_rate_limited = false;
+                        self.requests_sent.inc(Status::Failure);
+                        debug!(?err, ?peer, "send failed");
+                        self.update_performance(&peer, self.timeout);
+                        continue;
+                    }
+                }
             }
-            // If the message was sent to someone, add the request to the map
-            Ok(()) => {
-                self.active.insert(id, key);
+
+            // If this untargeted key had all peers rate-limited, no other key will
+            // succeed either (rate limits are per-peer), so stop early
+            if is_untargeted && all_rate_limited {
+                break;
             }
         }
+
+        // No keys could be fetched, set waiter to retry as soon as possible
+        self.waiter = earliest_next_try
+            .or_else(|| Some(self.context.current().saturating_add(self.retry_timeout)))
```

Comment:
use the time returned from check!

Reply:
Rather than setting a short backoff, went back to the (effectively) infinite backoff. Using `retry_timeout` works if no peers were rate-limited and all requests fail, but we shouldn't need to re-attempt.